### PR TITLE
Enable deployment to both Vercel and Fly.io

### PR DESCRIPTION
**Pull request by Josehower**
https://next-js-example-spring-2023-vienna-austria-yy9g.vercel.app/

### CODE REVIEW

File: database/connect.ts [hunk 1]
```diff
@@ -1,50 +1,28 @@
-import 'server-only';
-import { config } from 'dotenv-safe';
-import { headers } from 'next/headers';
-import postgres, { Sql } from 'postgres';
+import postgres from 'postgres';
+import { setEnvironmentVariables } from '../ley.config.mjs';
 
-if (!process.env.FLY_IO) config();
+// This loads all environment variables from a .env file
+// for all code after this line
```
Comment (karlhorky) [inline database/connect.ts]: we probably don't need this comment

File: database/connect.ts [hunk 2]
```diff
@@ -1,50 +1,28 @@
...
+setEnvironmentVariables();
 
+// Type needed for the connection function below
 declare module globalThis {
-  let postgresSqlClient: Sql;
+  let postgresSqlClient: ReturnType<typeof postgres> | undefined;
```
Comment (karlhorky) [inline database/connect.ts]: I think `Sql` is the new version, not `ReturnType<...>`

File: database/connect.ts [hunk 3]
```diff
@@ -1,50 +1,28 @@
...
 }
 
 // Connect only once to the database
 // https://github.com/vercel/next.js/issues/7811#issuecomment-715259370
 function connectOneTimeToDatabase() {
-  if (!('postgresSqlClient' in globalThis)) {
+  if (!globalThis.postgresSqlClient) {
```
Comment (karlhorky) [inline database/connect.ts]: this is also the old version

File: database/connect.ts [hunk 4]
```diff
@@ -1,50 +1,28 @@
...
-  // Ideally there would something built into Next.js for this,
-  // which has been requested here:
-  //
-  // https://github.com/vercel/next.js/discussions/50695
-  return ((
-    ...sqlParameters: Parameters<typeof globalThis.postgresSqlClient>
-  ) => {
-    headers();
-    return globalThis.postgresSqlClient(...sqlParameters);
-  }) as typeof globalThis.postgresSqlClient;
```
Comment (karlhorky) [inline database/connect.ts]: why is this removed?

File: databaseExample.mjs [hunk 5]
```diff
@@ -1,7 +1,9 @@
-import { config } from 'dotenv-safe';
 import postgres from 'postgres';
+import { setEnvironmentVariables } from './ley.config.mjs';
 
-config();
+// This loads all environment variables from a .env file
+// for all code after this line
```
Comment (karlhorky) [inline databaseExample.mjs]: we probably don't need this comment

File: ley.config.mjs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  if (unconfiguredEnvVars.length > 0) {
+    throw new Error(
+      `.env.example environment ${
+        unconfiguredEnvVars.length > 1 ? 'variables' : 'variable'
+      } ${unconfiguredEnvVars.join(', ')} not configured in .env file`,
+    );
+  }
+}
+
+setEnvironmentVariables();
```
Comment (karlhorky) [inline ley.config.mjs:3]: interesting, just realized that the side effect of this is that when we import the file, it also runs this function... maybe not a problem, but we should decide what we want to do with this

File: database/connect.ts [hunk 7]
```diff
@@ -1,50 +1,28 @@
...
-import postgres, { Sql } from 'postgres';
+import postgres from 'postgres';
+import { setEnvironmentVariables } from '../ley.config.mjs';
 
-if (!process.env.FLY_IO) config();
+// This loads all environment variables from a .env file
+// for all code after this line
+setEnvironmentVariables();
 
+// Type needed for the connection function below
```
Comment (karlhorky) [inline database/connect.ts]:
```suggestion
// Type needed for assigning `globalThis.postgresSqlClient` in `connectOneTimeToDatabase`
```

File: ley.config.mjs (hunk 6 above)
Comment (Josehower) [reply ley.config.mjs:3]: ok so you mean this function is going to run twice in the database connect file? 

still, I can't see why this may be an issue. But I am also open to using a different pattern then

Comment (karlhorky) [reply ley.config.mjs:3]: yeah we shouldn't run it twice - see if you can come up with a proposal or 2

**Comment by karlhorky** [comment]
The PR description should contain the deployed Vercel URL, after this has been deployed

File: ley.config.mjs (hunk 6 above)
Comment (Josehower) [reply ley.config.mjs:3]: Options:

1. Moving this to utils/config.mjs as we had thought on Friday
2. Separate the logic so only dotenv is being called on import 

```js
import fs from 'node:fs';
import dotenv from 'dotenv';

function envConfig() {

  dotenv.config();
  ... rest of the file check
}

export function setEnvironmentVariables() {
  if (process.env.NODE_ENV === 'production' || process.env.CI) {
    // Set standard environment variables for Postgres.js from Vercel environment variables
    if (process.env.POSTGRES_URL) {
   .... env update
    }
    return;
  }
}

envConfig();

const options = {
  ssl: Boolean(process.env.POSTGRES_URL),
};

export default options;
```

Comment (karlhorky) [reply ley.config.mjs:3]: hmm, yeah option 2 probably wouldn't be great, makes the code hard to understand

other options:

3. change the import in `database/connect.ts` to `import '../ley.config.mjs'` without any named import, this will cause the function to run as a side effect
4. add a condition around the `setEnvironmentVariables()` in `ley.config.mjs` to prevent running a 2nd time - check for either:
   1. an environment variable that exists when Next.js is started
   2. an environment that was set by the `setEnvironmentVariables()` function
5. add some kind of boolean variable that checks whether the `setEnvironmentVariables()` function has already run, and then avoids running it a second time

but probably your option 1 seems best

Comment (Josehower) [reply ley.config.mjs:3]: [2a1f618](https://github.com/upleveled/next-js-example-spring-2023-vienna-austria/pull/8/commits/7178735af66f9d3a8d03732eb77901a8182fe4ee)

File: databaseExample.mjs (hunk 5 above)
Comment (Josehower) [reply databaseExample.mjs]: 895e591

File: database/connect.ts (hunk 4 above)
Comment (Josehower) [reply database/connect.ts]: 95152c6

File: database/connect.ts (hunk 3 above)
Comment (Josehower) [reply database/connect.ts]: 95152c6

File: database/connect.ts (hunk 2 above)
Comment (Josehower) [reply database/connect.ts]: 95152c6

File: database/connect.ts (hunk 1 above)
Comment (Josehower) [reply database/connect.ts]: 421fd77

File: util/config.mjs
Comment (karlhorky) [inline util/config.mjs:1]: copying from https://github.com/upleveled/security-vulnerability-examples-next-js-postgres/pull/96/files/659586f51956d605e745e0b26f1f2beaf6da23e7#r1246405464 in case we want to open an issue in `ley`:

---

we need `.mjs` extension here because Ley cannot figure out imports of `.js` files inside `ley.config.mjs` - it leads to it incorrectly using `require()` - maybe a bug in Ley:

```
 security-vulnerability-examples-next-js-postgres git:(multi-platform--deploy-setup) ✗ pnpm migrate up

> security-vulnerability-examples-next-js-postgres@0.1.0 migrate /Users/josehower/projects-upleveled/security-vulnerability-examples-next-js-postgres
> ley "up"

node:internal/errors:490
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/josehower/projects-upleveled/security-vulnerability-examples-next-js-postgres/ley.config.mjs not supported.
Instead change the require of /Users/josehower/projects-upleveled/security-vulnerability-examples-next-js-postgres/ley.config.mjs to a dynamic import() which is available in all CommonJS modules.
    at exports.load (/Users/josehower/projects-upleveled/security-vulnerability-examples-next-js-postgres/node_modules/.pnpm/ley@0.8.1/node_modules/ley/lib/util.js:62:10)
    at async /Users/josehower/projects-upleveled/security-vulnerability-examples-next-js-postgres/node_modules/.pnpm/ley@0.8.1/node_modules/ley/bin.js:11:13 {
  code: 'ERR_REQUIRE_ESM'
}

Node.js v20.2.0
 ELIFECYCLE  Command failed with exit code 1.
```

Comment (karlhorky) [reply util/config.mjs:1]: Just noticing that @Josehower used Node.js v20 in the output above (and remembering some problems with Node.js v20 with some other Luke Edwards libs), wonder if this is the problem... 🤔

**Review by karlhorky — APPROVED** [review]

File: util/config.mjs
Comment (karlhorky) [reply util/config.mjs:1]: Nope, not Node.js v18 - apparently a problem with Ley:

```bash
➜  next-js-example-spring-2023-vienna-austria git:(multi-platform-deploy-setup) ✗ pnpm migrate up

> next-js-example-spring-2023-vienna-austria@0.1.0 migrate /home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria
> ley --require tsm "up"

node:internal/errors:490
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_REQUIRE_ESM]: require() of ES Module /home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria/ley.config.mjs not supported.
Instead change the require of /home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria/ley.config.mjs to a dynamic import() which is available in all CommonJS modules.
    at exports.load (/home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria/node_modules/.pnpm/ley@0.8.1/node_modules/ley/lib/util.js:62:10)
    at async /home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria/node_modules/.pnpm/ley@0.8.1/node_modules/ley/bin.js:11:13 {
  code: 'ERR_REQUIRE_ESM'
}

Node.js v18.16.1
 ELIFECYCLE  Command failed with exit code 1.
```

Comment (Josehower) [reply util/config.mjs:1]:
```
➜  ~ node -v
v18.16.1
➜  ~ z spring 
➜  next-js-example-spring-2023-vienna-austria git:(multi-platform-deploy-setup) ✗ pnpm migrate up

> next-js-example-spring-2023-vienna-austria@0.1.0 migrate /home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria
> ley --require tsm "up"

node:internal/errors:490
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_REQUIRE_ESM]: require() of ES Module /home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria/ley.config.mjs not supported.
Instead change the require of /home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria/ley.config.mjs to a dynamic import() which is available in all CommonJS modules.
    at exports.load (/home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria/node_modules/.pnpm/ley@0.8.1/node_modules/ley/lib/util.js:62:10)
    at async /home/josehower/upleveled-projects/next-js-example-spring-2023-vienna-austria/node_modules/.pnpm/ley@0.8.1/node_modules/ley/bin.js:11:13 {
  code: 'ERR_REQUIRE_ESM'
}

Node.js v18.16.1
```

Comment (karlhorky) [reply util/config.mjs:1]: Opened an issue about this unusual behavior here:

- https://github.com/lukeed/ley/issues/35
